### ISSUE TITLE
DocFX unable to "compile" Oqtane client

### ISSUE THREAD

**Issue by GregStevenson** [oqtane/oqtane.docs#12]
Trying to compile the Octane 5.0 documents. Octane 5.0 compiles as a project just fine as well as the Oqtane Docs Builder solution once everything is updated to use .Net 8, but running the post build event **dotnet docfx** fails with the following error:

...\oqtane.framework\Oqtane.Client\Modules\ModuleBase.cs(40,16): error CS0246: The type or namespace name 'ModuleInstance' could not be found (are you missing a using directive or an assembly reference?)

Get same error running **dotnet docfx** in a command prompt window.

**Comment by iJungleboy** [comment]
I'll check in 2-3 days. 

Could be .net 8 incompatibility, could also be that you are on an incomplete commit...

**Comment by GregStevenson** [comment]
@iJungleboy Thanks.

**Comment by GregStevenson** [comment]
@iJungleboy, I'm just following up on this issue to see if you could reproduce. However, I know how hectic the holiday season can be. In the meantime, I've obtained used copies of the _Professional DNN7_ and _Professional DotNetNuke Module Programming_ books through Amazon to act as "Style Guides". Using them as a reference, I've been using GitBook to start writing custom module documentation in Markdown with the help of ChatGPT4 and Claude2.

**Comment by thabaum** [comment]
@GregStevenson I will take a look into this as well to see if I can help at all with the build issue shortly.  I want to be sure all the different ways I use or am going to use Oqtane are covered well so I can use them as a go to resource.  I am hoping some PRs produced just now help get the ball rolling for me here as a contributor.  I want to see what information from the Oqtane Blog can and should be included here in some form so we have a base for the current Oqtane Version 5 as a new foundation.  It would be nice to be sure every aspect of the framework is covered.  This should help uncover any issues we can attack testing all different scenarios for documenting Oqtane.

@iJungleboy I will throw down what I can when I can and feel I have something worthy for your time to review, I appreciate your help here.  I look forward to a blog/wiki/docfx solutions for all who wish to contribute to which ever documentation projects benefit them the most.  A piece here and another there eventually the puzzle gets filled.

I would like to review ways to keep things from being too redundant like in the installations doc page as it will get pretty long eventually.  However not a big deal I tried at the top to put a generic version then release version and why use a release version.  This all can be improved upon as I was just providing some insight on a potential solution with the PR #14  provided.

**Comment by GregStevenson** [comment]
@thabaum Thanks for looking into this and your other contributions to Oqtane.

**Comment by thabaum** [comment]
@GregStevenson I was able to make some progress I believe on the compilation issue for me being a first time ever trying and getting to know "docfx".

I updated the project to use "net8.0" in the project files.

That led to that same error still however I resolved some warnings to put in as some additional issues with docs.

So I found out it was relating to the project trying to run the command "dotnet docfx".

I then proceeded to open the terminal and run in manually which then told me a bit more information that I needed to run "dotnet restore tools" or installing missing docfx to get it to compile I believe.

I will create a PR maybe you can test see if it resolves.

Here is the updated project file ensure you have .NET 8 installed.

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Library</OutputType>
		<TargetFramework>net8.0</TargetFramework>
		<RestoreOnBuild>true</RestoreOnBuild>
	</PropertyGroup>
	<Target Name="PreBuild" BeforeTargets="PreBuildEvent">
		<Exec Command="dotnet tool list -g | findstr docfx || dotnet tool install --global docfx" />
	</Target>
	<Target Name="PostBuild" AfterTargets="PostBuildEvent">
		<Exec Command="dotnet docfx" />
	</Target>
</Project>
```

I have tested this solution out and has worked with success 2 out of 2 times!

**Comment by thabaum** [comment]
After further investigation of my issue I was able to unblock the file I got blocked somehow along the way.
I used the following command in terminal while navigated to the `.config` folder

To unblock the file I ran this:
`Unblock-File -Path .\dotnet-tools.json`

Next I ran the following command to restore the DocFX package:
`dotnet tool restore`

![image](https://github.com/oqtane/oqtane.docs/assets/13577556/c63af03b-2858-442b-83cb-3008b0b294c5)

After running those with the latest .NET 8 updated files I was able to restore docfx:
![image](https://github.com/oqtane/oqtane.docs/assets/13577556/f25e38c0-e13d-4ec3-87cf-aebb5652008f)

And ran a successful build.

![image](https://github.com/oqtane/oqtane.docs/assets/13577556/f6d41461-20e9-45eb-8bfd-1098db676020)

I hope this helps, cheers!

**Comment by GregStevenson** [comment]
@thabaum Thanks for looking into this. I've been under the weather for the past few days, but will try out your suggestions today. I tried a lot of the items you mentioned before posting the issue, but not Unblock-File. I'll let you know if it works for me as well.

**Comment by thabaum** [comment]
@GregStevenson I hope you are feeling better and are able to get Oqtane Docs ready to roll!  I was able to host this on IIS as well without an issue.  I am having issue with a Oqtane Framework instance on IIS however I need to iron out in my IDE still.  I put a few moments a day towards each, maybe today I can get that done up as well and start feeling out some fresh ideas.

It was the one thing that brought that error, I am not sure if we need to upload a new file that is not locked down with permissions from whomever uploaded it last?  I am hoping we can just upload a new file, however I copy that repo over after it is unblocked, and it is blocked again.  So I think it is a VS environment issue at least on my end.  Maybe if you open your VS instance up with Run As Administrator it works it out.

My repo also adds installing DocFX if it does not exist which was another issue I had.

Issue #15 should handle all the updates needed to files to run .NET 8 and latest version of DocFX.

**Comment by GregStevenson** [comment]
@thabaum I tried again without success. As I eluded to earlier, I had already done the things you did except the unlock which is not a problem for me. Oqtane.docs and the framework compile just fine for me. It is the post build event where I have issues and looking back, I did not post my exact errors. I'm getting this doing a **dotnet docfx** from command line.
![image](https://github.com/oqtane/oqtane.docs/assets/10802205/63479662-2e0e-4677-8f74-0bb9a9d7ddf2)
It's very weird that ModuleInstance cannot be found since everything compiles just fine.

**Comment by thabaum** [comment]
I am trying to understand why you have oqtane framework building with it?

Are you trying to create a site that runs the DocFX site?  You may need to modify Oqtane to do this?  

I just built the Oqtane Docs Builder solution and ran it as deploy I think or release to build the /docs folder outside of the project, used IIS to host that folder.

**Comment by thabaum** [comment]
@GregStevenson I would not mind getting some feedback on how this [PR repo](https://github.com/thabaum/oqtane.docs/tree/update-net8-install-docfx-prebuild) works for you.  It updates everything to run the latest versions of .NET  and DocFX along with related doc updates.

My workflow is this:

Download latest source for latest [Oqtane release](https://github.com/oqtane/oqtane.framework/releases) and unzip it to a folder along with this repo placed in the same folder `oqtane.docs` and `oqtane.framework`.

![image](https://github.com/oqtane/oqtane.docs/assets/13577556/99bfd510-40b7-413c-b782-ca02160a8555)

If I am understanding things more clearly now as I am green here as well... When you build DocFX it will scan documentation comments created in the code files located in the `oqtane.framework` folder and  thencompile them into the `oqtane.docs/docs` folder which it's entire contents is what you are able to host as an IIS website.

Maybe I am doing this wrong, I will need to review I am not missing any api docs...  I am trying to troubleshoot why I cannot run Oqtane itself on IIS using Windows 11 at the same time as helping support issues with docs :).  I can run docs, but not Oqtane which is odd.  I may test out a few alternatives so correct me anywhere I can use it.

I hope this helps, cheers!

**Comment by GregStevenson** [comment]
AFAIK, I just cloned the Oqtane.docs repository in the same folder that contains Oqtane.framework since that is how the bulk of the Oqtane repos seem to work. My guess is that it is looking at the framework for code comments to build API docs. I did nothing special other than updating target framework to .Net8 and similar minor changes. I'm in the middle of something now, but I'll try to run your repo by tomorrow noon PST.

**Comment by thabaum** [comment]
When you test make sure the file structure looks similar to:

![image](https://github.com/oqtane/oqtane.docs/assets/13577556/d8ff2ccd-61a3-46d0-bd01-0e0fe3bcbf5a)

![image](https://github.com/oqtane/oqtane.docs/assets/13577556/786edbf1-01bd-4bb3-8b45-ac777995b663)


`oqtane.docs\Oqtane.Docs`
`oqtane.framework\Oqtane.Server`

Both of these should be found in the root folder used to put the two projects and named/renamed exactly the same for example.

The docs site will compile in `oqtane.docs\docs` folder

I downloaded the zip files and made sure they did this.

Then I ran the solution inside the oqtane.docs folder that contains the Oqtane.Docs project.

You may have all the solutions/projects in the same folder which would not be correct for the way I was able to compile the docs I do not believe.

I have a blank page for API pretty much on release some reason, not sure why.  All the API files are there but the tree is missing to navigate them:
current build:
![image](https://github.com/oqtane/oqtane.docs/assets/13577556/1c944ef8-7984-4acf-bb03-cfe29fad5f26)

docs.oqtane.org:
![image](https://github.com/oqtane/oqtane.docs/assets/13577556/cf96ae90-cff6-443c-b24b-de1245143e04)

maybe i am doing something wrong or something is missing or not enabled to allow this to work i will look closer at this later.

**Comment by iJungleboy** [comment]
@thabaum as I mentioned in the pull request #15 solves all the .net 8 problems, but the menu in the API is missing (just like you noticed as well).

I'll try 2-3 things, but we may have to switch to another theme, like I did on https://docs.2sxc.org

**Comment by iJungleboy** [comment]
I think I have it figured out...

various breaking docfx changes...

**Comment by iJungleboy** [comment]
just fixed everything and pushed the latest build.

should be on https://docs.oqtane.org in ca. 3 mins

**Comment by GregStevenson** [comment]
@iJungleboy I just cloned and compiled the oqtane.docs repository. After a **dotnet tool restore**,  everything compiled as expected but when opening the output, _file:///<path to oqtane>/oqtane.docs/docs/articles/documentation/index.html_ locally in browser, I still experience, not menu, theming, icons, etc as was described earlier. You last comment mentioned this should fix everything. What am I missing?

**Comment by iJungleboy** [comment]
@GregStevenson I actually assume that you started with the wrong instructions, resulting in this.

I will make sure there are only one set of instructions and enhance them to explain both Visual Studio and VS Code use. 

#35 #36

Should be out in ca. 1h

**Comment by iJungleboy** [comment]
Just push very updated docs.

My guess is you didn't have the `oqtane.framework` in the right place, so the compiler couldn't find it - and not generate the API docs. Check out the updated instructions: https://docs.oqtane.org/guides/docs/index.html and give me feedback ;)

**Comment by GregStevenson** [comment]
No that is not it. oqtane.framework and oqtane.docs are siblings. Everything compiled and the api files are present as shown. In fact, viewing one it seems to find the styles, etc.
![image](https://github.com/oqtane/oqtane.docs/assets/10802205/6945519a-a282-4223-92c8-1a837f8c0bac)
Maybe there is some issue launching the index.html page locally rather than using a webserver. AFAICT all the files were generated and in their right relative places in the docs folder, but the index page has no styling and no menus on top and left side.

![image](https://github.com/oqtane/oqtane.docs/assets/10802205/078be71e-8a5f-4742-94e6-a377dd8edceb)

I did a compare of the view source of what was generated and what is at [octane.org](https://docs.oqtane.org/guides/docs/index.html ). There are a lot of differences. Including the fact that my locally rendered version references the fact that it is 5.01. Indicating it would be current.

![image](https://github.com/oqtane/oqtane.docs/assets/10802205/3eff59d3-761d-47fc-9cfd-609544a53849)

**Comment by iJungleboy** [comment]
@GregStevenson To me it looks as if you're browsing the html files from your local disk.

You need to run it through a web server, otherwise most things don't look right.

It's explained in the docs, incl. how to set it up in VS Code with a live web server, or in IIS.

https://docs.oqtane.org/guides/docs/build.html#build-and-test-locally-using-vs-code

https://docs.oqtane.org/guides/docs/build.html#build-and-test-locally-using-visual-studio-2022

https://docs.oqtane.org/guides/docs/run-in-iis.html

**Comment by GregStevenson** [comment]
Ok, I'm good now. Used GoLive in VSCode. Somehow I missed this:
The generated docs are static pages and contain various JavaScripts to provide menu and search functionality. **These do not work, if you just open the index.html in a browser.** So you need to run it in a local IIS to fully test it.
